crates/vm/src/arch/segment.rs

```diff
@@ -157,12 +79,42 @@ where
     pub metrics: VmMetrics,
 }
 
-pub struct ExecutionSegmentState {
-    pub pc: u32,
-    pub is_terminated: bool,
+pub struct TracegenCtx {
+    pub timestamp: u32,
 }
 
-impl<F: PrimeField32, VC: VmConfig<F>> ExecutionSegment<F, VC> {
+impl TracegenCtx {
+    pub fn new(timestamp: u32) -> Self {
+        Self { timestamp }
+    }
+}
+
+pub type TracegenVmExecutionState = VmExecutionState<AddressMap<PAGE_SIZE>, TracegenCtx>;
+
+impl TracegenVmExecutionState {
+    pub fn from_pc_and_memory_controller<F>(
+        pc: u32,
+        memory_controller: &MemoryController<F>,
+    ) -> Self
+    where
+        F: PrimeField32,
+    {
+        let memory = memory_controller.memory_image().clone();
```

Comment:
this is pretty bad to clone? is this temporary?

Reply:
yep, temporary. i need to figure out how to get memory here since it is owned by the memory controller atm